umbral-pre/src/bindings_wasm.rs

```diff
@@ -0,0 +1,624 @@
+extern crate alloc;
+
+use alloc::boxed::Box;
+use alloc::format;
+use alloc::string::String;
+use alloc::{vec, vec::Vec};
+use core::fmt;
+
+use js_sys::Error;
+use wasm_bindgen::prelude::{wasm_bindgen, JsValue};
+
+use serde::{Deserialize, Serialize};
+
+use crate as umbral_pre;
+use crate::{DeserializableFromArray, SerializableToArray, SerializableToSecretArray};
+
+fn map_js_err<T: fmt::Display>(err: T) -> JsValue {
+    Error::new(&format!("{}", err)).into()
+}
+
+#[wasm_bindgen]
+pub struct SecretKey {
+    backend: umbral_pre::SecretKey,
+}
+
+#[wasm_bindgen]
+impl SecretKey {
+    /// Generates a secret key using the default RNG and returns it.
+    pub fn random() -> Self {
+        Self {
+            backend: umbral_pre::SecretKey::random(),
+        }
+    }
+
+    /// Generates a secret key using the default RNG and returns it.
+    #[wasm_bindgen(js_name = publicKey)]
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey {
+            backend: self.backend.public_key(),
+        }
+    }
+
+    #[wasm_bindgen(js_name = toSecretBytes)]
+    pub fn to_secret_bytes(&self) -> Box<[u8]> {
+        self.backend
+            .to_secret_array()
+            .as_secret()
+            .to_vec()
+            .into_boxed_slice()
+    }
+
+    #[wasm_bindgen(js_name = fromBytes)]
+    pub fn from_bytes(data: &[u8]) -> Result<SecretKey, JsValue> {
+        umbral_pre::SecretKey::from_bytes(data)
+            .map(|sk| Self { backend: sk })
+            .map_err(map_js_err)
+    }
+
+    #[allow(clippy::inherent_to_string)]
+    #[wasm_bindgen(js_name = toString)]
+    pub fn to_string(&self) -> String {
+        format!("{}", self.backend)
+    }
+}
+
+impl SecretKey {
+    pub fn inner(&self) -> &umbral_pre::SecretKey {
```

Comment:
Is that for use in `nucypher_core`? Perhaps it can be done similarly to the Python bindings (`AsBackend` and `FromBackend` traits)

Reply:
I will try this out, thanks!